### PR TITLE
EVG-3567: create job to remove image so disk does not fill up

### PR DESCRIPTION
**Pull request by li-cindy**
Removes earliest created image on a host when docker images start taking up a lot of disk space (images estimated to be ~11GB, currently allowing 10 images at a time). Tested on a host I spawned as well.

### CODE REVIEW

File: cloud/docker.go [hunk 1]
```diff
@@ -249,3 +249,42 @@ func (m *dockerManager) GetContainers(ctx context.Context, h *host.Host) ([]stri
...
 	return ids, nil
 }
+
+func (m *dockerManager) RemoveLeastRecentlyUsedImageID(ctx context.Context, h *host.Host) error {
+	images, err := m.client.ListImages(ctx, h)
+	if err != nil {
+		return errors.Wrap(err, "Error listing images")
+	}
+	// get earliest created image ID
+	earliest := time.Now()
```
Comment (john-m-liu) [inline cloud/docker.go]: You aren't removing the least recently used image - you're removing the oldest image. I don't think that's a problem necessarily, but you should either implement LRU or rename this to oldest

File: cloud/docker.go [hunk 2]
```diff
@@ -249,3 +249,42 @@ func (m *dockerManager) GetContainers(ctx context.Context, h *host.Host) ([]stri
...
+			earliest = time.Unix(image.Created, 0)
+			id = image.ID
+		}
+	}
+
+	if id == "" {
+		return errors.Errorf("No images to delete on host '%s'", h.Id)
+	}
+	// remove image based on ID
+	err = m.client.RemoveImage(ctx, h, id)
```
Comment (john-m-liu) [inline cloud/docker.go]: Do we need to validate that there are no containers running this image before removing?

File: cloud/docker.go [hunk 3]
```diff
@@ -249,3 +249,42 @@ func (m *dockerManager) GetContainers(ctx context.Context, h *host.Host) ([]stri
...
+	return nil
+}
+
+func (m *dockerManager) CalculateImageSpaceUsage(ctx context.Context, h *host.Host) (int64, error) {
+	images, err := m.client.ListImages(ctx, h)
+	if err != nil {
+		return 0, errors.Wrap(err, "Error listing images")
+	}
+
+	space := int64(0)
```
Comment (john-m-liu) [inline cloud/docker.go]: This should be named with its units and the function should have a comment with its units as well

File: cloud/docker_client.go [hunk 4]
```diff
@@ -266,6 +268,46 @@ func (c *dockerClientImpl) ListContainers(ctx context.Context, h *host.Host) ([]
...
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to generate docker client")
+	}
+
+	// Get all container images
+	opts := types.ImageListOptions{All: false}
+	images, err := dockerClient.ImageList(ctx, opts)
+	if err != nil {
+		err = errors.Wrap(err, "Docker list API call failed")
+		grip.Error(err)
```
Comment (john-m-liu) [inline cloud/docker_client.go]: You don't need to both log and return the error - I would just return the error

File: units/lru_container_removal.go [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+	j.HostID = h.Id
+
+	j.SetID(fmt.Sprintf("%s.%s", leastRecentlyUsedImageJobName, id))
+	return j
+}
+
+func (j *leastRecentlyUsedImageJob) Run(ctx context.Context) {
+	var cancel context.CancelFunc
+
+	ctx, cancel = context.WithCancel(ctx)
```
Comment (john-m-liu) [inline units/lru_container_removal.go]: There's no need to add a canceler here unless you're going to explicitly cancel the context at some point

**Comment by li-cindy** [comment]
@john-m-liu I added some changes as per your comments. I check to see if there are containers on an image, and if there are, I delete them. However, I'm wondering if there are some bad side effects of this (we probably want to keep these containers around if they are running?). Let me know if I should actually be checking for the oldest image with **no** containers on them.

File: cloud/docker.go [hunk 6]
```diff
@@ -249,3 +249,78 @@ func (m *dockerManager) GetContainers(ctx context.Context, h *host.Host) ([]stri
...
+
+	// remove containers on image before removing the image
+	containersRunningImage, err := m.GetContainersRunningImage(ctx, h, id)
+	if err != nil {
+		return errors.Wrapf(err, "Error getting containers running on image '%s'", id)
+	}
+
+	if len(containersRunningImage) > 0 {
+		for _, containerID := range containersRunningImage {
+			if err := m.client.RemoveContainer(ctx, h, containerID); err != nil {
```
Comment (john-m-liu) [inline cloud/docker.go]: Instead of removing the container, let's continue onto the next oldest

File: cloud/docker.go [hunk 7]
```diff
@@ -249,3 +249,78 @@ func (m *dockerManager) GetContainers(ctx context.Context, h *host.Host) ([]stri
 
 	return ids, nil
 }
+
+// GetContainersRunningImage returns all the containers that are running a particular image
+func (m *dockerManager) GetContainersRunningImage(ctx context.Context, h *host.Host, imageID string) ([]string, error) {
```
Comment (john-m-liu) [inline cloud/docker.go]: This can be private

File: cloud/docker_client.go [hunk 8]
```diff
@@ -266,6 +268,45 @@ func (c *dockerClientImpl) ListContainers(ctx context.Context, h *host.Host) ([]
...
+}
+
+// RemoveImage forcibly removes an image from its host machine
+func (c *dockerClientImpl) RemoveImage(ctx context.Context, h *host.Host, imageID string) error {
+	dockerClient, err := c.generateClient(h)
+	if err != nil {
+		return errors.Wrap(err, "Failed to generate docker client")
+	}
+
+	opts := types.ImageRemoveOptions{Force: true}
```
Comment (john-m-liu) [inline cloud/docker_client.go]: What's the difference if we don't pass the force param?

File: cloud/docker_client.go [hunk 9]
```diff
@@ -266,6 +268,45 @@ func (c *dockerClientImpl) ListContainers(ctx context.Context, h *host.Host) ([]
...
+	removed, err := dockerClient.ImageRemove(ctx, imageID, opts)
+	if err != nil {
+		err = errors.Wrapf(err, "Failed to remove image '%s'", imageID)
+		grip.Error(err)
```
Comment (john-m-liu) [inline cloud/docker_client.go]: There's no need to both log and return an error - I would just return it

File: units/oldest_container_removal.go [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+	"github.com/evergreen-ci/evergreen/model/host"
+	"github.com/mongodb/amboy"
+	"github.com/mongodb/amboy/dependency"
+	"github.com/mongodb/amboy/job"
+	"github.com/mongodb/amboy/registry"
+	"github.com/pkg/errors"
+)
+
+const (
+	oldestImageJobName = "oldest-image"
```
Comment (john-m-liu) [inline units/oldest_container_removal.go]: I feel like this needs a verb, maybe "oldest-image-removal." Also the file name is inconsistent with this

File: cloud/docker_client.go (hunk 8 above)
Comment (li-cindy) [reply cloud/docker_client.go]: Even if all containers running this image have been stopped, leaving out the force param results in this error: 
`Failed to remove image 'sha256:74f8760a2a8b28abade3fcbcdb6998543f1d9b4a6fb61463c10adc0765c3cb12': Error response from daemon: {"message":"conflict: unable to delete 74f8760a2a8b (must be forced) - image is being used by stopped container ba39421ec78d"}`

File: cloud/docker.go (hunk 6 above)
Comment (li-cindy) [reply cloud/docker.go]: Done.

**Pull request merged**